rclpy/rclpy/parameter_event_handler.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from itertools import chain
+from multiprocessing import Lock
+from typing import Callable, Dict, Iterable, List, Optional, Tuple
+
+
+from rcl_interfaces.msg import ParameterEvent
+from rclpy.callback_groups import CallbackGroup
+from rclpy.event_handler import SubscriptionEventCallbacks
+from rclpy.node import Node
+from rclpy.parameter import Parameter
+from rclpy.qos import qos_profile_parameter_events
+from rclpy.qos import QoSProfile
+from rclpy.qos_overriding_options import QoSOverridingOptions
+
+
+class ParameterCallbackHandle:
+    def __init__(
+        self,
+        parameter_name: str,
+        node_name: str,
+        callback: Callable[[Parameter], None],
+    ):
+        self.parameter_name = parameter_name
+        self.node_name = node_name
+        self.callback = callback
+
+        self.mutex = Lock()
+
+
+class ParameterEventCallbackHandle:
+
+    def __init__(self, callback: Callable[[ParameterEvent], None]):
+        self.callback = callback
+        self.mutex = Lock()
+
+
+class ParameterEventHandler:
+    class Callbacks:
+
+        def __init__(self):
+            self.parameter_callbacks: Dict[
+                Tuple[str, str], List[ParameterCallbackHandle]
+            ] = defaultdict(list)
+            self.event_callbacks: List[ParameterEventCallbackHandle] = []
+            self.mutex = Lock()
+
+        def event_callback(self, event: ParameterEvent):
+            with self.mutex:
+                for (param_name, node_name), callbacks_list in self.parameter_callbacks.items():
+                    if parameter := ParameterEventHandler.get_parameter_from_event(
+                        event, parameter_name=param_name, node_name=node_name
+                    ):
+                        for callback_handle in reversed(callbacks_list):
```

Comment:
Given that this is the "hot" loop, I think it would actually probably be better to put the callbacks into the list in the proper order at insertion time.  That is, rather than doing a `callbacks_list.append()` and then reversing the list here, we should probably use `callbacks_list.insert(0, handle)` below, and just iterate forwards through the list here.

The same goes for the call to `reversed` below.